File: src/main/java/org/app/infrastructure/local/InMemoryRelationshipRepository.java
package org.app.infrastructure.local;

import jakarta.enterprise.context.ApplicationScoped;
import org.app.domain.IRelationshipRepository;
import org.app.domain.Relationship;

import java.util.*;
import java.util.stream.Collectors;

@ApplicationScoped
public class InMemoryRelationshipRepository implements IRelationshipRepository {
    List<RelationshipData> relationships;

    public InMemoryRelationshipRepository() {
        this.relationships = new ArrayList<>();
    }

    private boolean matchFollowingAndFanWithRelationship(RelationshipData relationship, UUID followingId, UUID fanId){
        return relationship.getFollowingId().equals(followingId.toString()) && relationship.getFanId().equals(fanId.toString()) ||
                relationship.getFollowingId().equals(fanId.toString()) && relationship.getFanId().equals(followingId.toString());
    }

    private List<Relationship> toRelationships(List<RelationshipData> relationshipData){
        return relationshipData.stream().map(RelationshipData::toDomain).collect(Collectors.toList());
    }

    private List<RelationshipData> toRelationshipsData(List<Relationship> relationships){
        return relationships.stream().map(RelationshipData::toData).collect(Collectors.toList());
    }

    @Override
    public List<Relationship> find(UUID followingId, UUID fanId) {
        List<RelationshipData> relationshipData = relationships.stream().filter(relationship -> matchFollowingAndFanWithRelationship(relationship, followingId, fanId)).collect(Collectors.toList());
        return toRelationships(relationshipData);
    }

    private boolean matchRelationship(RelationshipData relationship, String followingId, String fanId){
        return relationship.getFollowingId().equals(followingId) && relationship.getFanId().equals(fanId);
    }

    @Override
    public List<Relationship> findByFollowingId(UUID followingId, UUID fanId) {
        List<RelationshipData> relationshipData = relationships.stream().filter(relationship -> matchRelationship(relationship, followingId.toString(), fanId.toString())).collect(Collectors.toList());
        return toRelationships(relationshipData);
    }

    private boolean matchFollowingOrFanWithRelationship(RelationshipData relationship, UUID userId){
        return relationship.getFollowingId().equals(userId.toString()) || relationship.getFanId().equals(userId.toString());
    }

    @Override
    public List<Relationship> findByUserId(UUID userId) {
        List<RelationshipData> relationshipData = relationships.stream().filter(relationship -> matchFollowingOrFanWithRelationship(relationship, userId)).collect(Collectors.toList());
        return toRelationships(relationshipData);
    }

    @Override
    public void addAll(List<Relationship> relationships) {
        this.relationships.addAll(toRelationshipsData(relationships));
    }

    private Optional<RelationshipData> getRelationship(List<RelationshipData> relationshipsData, RelationshipData targetRelationship){
        return relationshipsData.stream().filter(relationship -> matchRelationship(relationship, targetRelationship.getFollowingId(), targetRelationship.getFanId())).findFirst();
    }

    @Override
    public void update(List<Relationship> relationships) {
        this.relationships = this.relationships.stream().map(relationship -> getRelationship(toRelationshipsData(relationships), relationship).orElse(relationship)).collect(Collectors.toList());
    }

    @Override
    public void remove(List<Relationship> relationships) {
        this.relationships = this.relationships.stream().filter(relationship -> getRelationship(toRelationshipsData(relationships), relationship).isEmpty()).collect(Collectors.toList());
    }
}

File: go/app/main/infrastructure/local/in_memory_relationship_repository.go
package local

import (
    "github.com/google/uuid"
    "mutualAttentionSystem/app/main/domain"
)

type InMemoryRelationshipRepository struct {
    relationships []*RelationshipData
}

func NewInMemoryRelationshipRepository() *InMemoryRelationshipRepository {
    return &InMemoryRelationshipRepository{}
}

func filterFollowingOrFanWithRelationship(relationshipData *[]*RelationshipData, relationships []*RelationshipData, userId string) {
    for _, relationship := range relationships {
        if relationship.FanId == userId || relationship.FollowingId == userId {
            *relationshipData = append(*relationshipData, relationship)
        }
    }
}

func toRelationshipsDomain(relationshipsData []*RelationshipData) []*domain.Relationship {
    var relationship []*domain.Relationship
    for _, relationshipData := range relationshipsData {
        relationship = append(relationship, toRelationshipDomain(relationshipData))
    }
    return relationship
}

func (repository *InMemoryRelationshipRepository) FindUserId(id uuid.UUID) []*domain.Relationship {
    var relationshipsData []*RelationshipData
    filterFollowingOrFanWithRelationship(&relationshipsData, repository.relationships, id.String())
    return toRelationshipsDomain(relationshipsData)
}

func filterFollowingAndFanWithRelationship(relationshipData *[]*RelationshipData, relationships []*RelationshipData, followingId string, fanId string) {
    for _, relationship := range relationships {
        if relationship.FollowingId == followingId && relationship.FanId == fanId ||
            relationship.FollowingId == fanId && relationship.FanId == followingId {
            *relationshipData = append(*relationshipData, relationship)
        }

        if len(*relationshipData) == 2 {
            return
        }
    }
}

func (repository *InMemoryRelationshipRepository) Find(followingId uuid.UUID, fanId uuid.UUID) []*domain.Relationship {
    var relationshipsData []*RelationshipData
    filterFollowingAndFanWithRelationship(&relationshipsData, repository.relationships, followingId.String(), fanId.String())
    return toRelationshipsDomain(relationshipsData)
}

func toRelationshipsData(relationships []*domain.Relationship) []*RelationshipData {
    var relationshipsData []*RelationshipData
    for _, relationship := range relationships {
        relationshipsData = append(relationshipsData, toRelationshipData(relationship))
    }
    return relationshipsData
}

func filterRelationshipById(selfRelatinoships []*RelationshipData, relationshipId string) int {
    for index, relationship := range selfRelatinoships {
        if relationship.Id == relationshipId {
            return index
        }
    }
    return -1
}

func (repository *InMemoryRelationshipRepository) Update(relationships []*domain.Relationship) {
    relationshipsData := toRelationshipsData(relationships)
    for _, relationshipData := range relationshipsData {
        index := filterRelationshipById(repository.relationships, relationshipData.Id)
        if index == -1 {
            repository.relationships = append(repository.relationships, relationshipData)
        } else {
            repository.relationships[index] = relationshipData
        }
    }
}

func (repository *InMemoryRelationshipRepository) Remove(followingId uuid.UUID, fanId uuid.UUID) {
    for index, relationship := range repository.relationships {
        if relationship.FollowingId == followingId.String() && relationship.FanId == fanId.String() {
            repository.relationships = append(repository.relationships[:index], repository.relationships[index+1:]...)
        }
    }
}

func (repository *InMemoryRelationshipRepository) Save(relationships []*domain.Relationship) {
    relationshipsData := toRelationshipsData(relationships)
    repository.relationships = append(repository.relationships, relationshipsData...)
}
